### PR TITLE
Port gateway extension

### PR DESCRIPTION
**Pull request by neonjonn**
If possible, I'd like to keep this PR open while I attempt to add welcome cards as attachments to the welcome messages

### CODE REVIEW

**Comment by parafoxia** [comment]
It looks like you might to rebase the commit chain. If you do the following commands, it *should* fix it:
```sh
git remote add upstream https://github.com/Carberra/Carberretta
git pull upstream main
git rebase
```

**Comment by parafoxia** [comment]
Also, I'd prefer if you put the attachment images in a separate PR as the scopes are different. Though you can obvs just attach it to the code you wrote here. If you've done everything for porting, mark it as ready for review and I'll have a look-see.

**Comment by neonjonn** [comment]
I'm still getting used to making PRs 😅

Did the rebase, never knew about that command before :0

And cool, I'll make another PR after adding the attachment code :)

**Comment by parafoxia** [comment]
Yeah no worries haha, they can be a nightmare to do sometimes haha. It doesn't look like it did what I thought it might do, so I'll have a look and see if I can fix up the history a bit. You also need to fix the typing issue that's arising.

**Comment by neonjonn** [comment]
I forgot to sync the commit, will do now so hopefully that fixes the rebase. And I almost never use mypy but I'll take a look at the typing issues now

**Comment by parafoxia** [comment]
Damn, might have just messed up it completely lmao.

**Comment by neonjonn** [comment]
Oh no...

**Comment by parafoxia** [comment]
Okay that fixed it by the looks of it. Literally had to manually rebase about 7 commits but got there in the end lmao.

**Comment by neonjonn** [comment]
Oh damn, well at least it's fixed now 😅

**Comment by parafoxia** [comment]
Yeah 😆 I think it may have been my fault cos I did do a few rebases yesterday, so I think your downstream fork got caught up in that. To fix you downstream main though, you'll need to do:
```sh
git checkout main
git reset --hard HEAD~10
git pull upstream main
```
That won't affect your other branch.

**Comment by neonjonn** [comment]
yeah that seemed to work, all looks good

**Comment by parafoxia** [comment]
Sweet

File: carberretta/extensions/gateway.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+async def on_member_update(event: hikari.MemberUpdateEvent) -> None:
+    if not event.old_member or not event.member:
+        return
+
+    if event.old_member.is_pending != event.member.is_pending:
+        humans = len(
+            [
+                m
+                async for m in plugin.bot.rest.fetch_members(Config.GUILD_ID)
+                if not m.bot
```
Comment (parafoxia) [inline carberretta/extensions/gateway.py:120]: This should be `m.is_bot`.

**Review by parafoxia — CHANGES_REQUESTED** [review]
The typing seems to be okay now Lightbulb has updated, but you've made a typo that needs to be fixed. I'm on mobile atm but when I'm next on my computer I'll pull the changes down and test everything properly.

**Comment by parafoxia** [comment]
Also if you can change Lightbulb's min version from 2.0 to 2.0.2 that'd be ace.

**Comment by neonjonn** [comment]
Cool, I'll make those changes now... and stop using GitHub Co-pilot, that's probably why I made that typo 😅

**Comment by parafoxia** [comment]
Tfw Copilot doesn't even get it right. Probably used Carberretta's old code to get that result lmao.

**Comment by neonjonn** [comment]
Probably XD, I had commented out the old code to reference it

File: carberretta/extensions/gateway.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    try:
+        plugin.bot.d.scheduler.get_job(f"{member.id}").remove()
+        return
+    except AttributeError:
+        if member.is_pending:
+            return
+
+        await plugin.bot.rest.create_message(
+            Config.GATEWAY_ID,
```
Comment (parafoxia) [inline carberretta/extensions/gateway.py:105]:
```suggestion
            Config.GATEWAY_CHANNEL_ID,
```
Change to conform with new channel env standard.

File: carberretta/extensions/gateway.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    if event.old_member.is_pending != event.member.is_pending:
+        humans = len(
+            [
+                m
+                async for m in plugin.bot.rest.fetch_members(Config.GUILD_ID)
+                if not m.is_bot
+            ]
+        )
+        await plugin.bot.rest.create_message(
+            Config.GATEWAY_ID,
```
Comment (parafoxia) [inline carberretta/extensions/gateway.py:124]:
```suggestion
            Config.GATEWAY_CHANNEL_ID,
```
Change to conform with new channel env standard.

File: carberretta/extensions/gateway.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        id=f"{member.id}",
+        next_run_time=dt.datetime.utcnow() + dt.timedelta(seconds=secs),
+        args=[member],
+    )
+
+
+@plugin.listener(hikari.StartedEvent)
+async def on_started(_: hikari.StartedEvent) -> None:
+    async for m in plugin.bot.rest.fetch_members(Config.GUILD_ID):
+        if (secs := (dt.datetime.utcnow() - m.joined_at).seconds) <= TIMEOUT:
```
Comment (parafoxia) [inline carberretta/extensions/gateway.py:74]:
```suggestion
    now = dt.datetime.now().astimezone()
    
    async for m in plugin.bot.rest.fetch_members(Config.GUILD_ID):
        if (secs := (now - m.joined_at).seconds) <= TIMEOUT:
```
This needs to be changed as Hikari uses offset-aware datetime objects, unlike discord.py.

**Review by parafoxia — CHANGES_REQUESTED** [review]
Haven't been able to do a proper test of the logic yet, but I think it should be okay after these changes. There are some differences between Hikari and discord.py that you need to address.

I also ask if you could port over the [patch introduced in v1.5.5](https://github.com/Carberra/Carberretta/commit/412b7269b8c4ed3872da883349f49f1f38f9f732), which fixes a problem related to cross-server gateway messages.

File: carberretta/extensions/gateway.py (hunk 2 above)
Comment (neonjonn) [reply carberretta/extensions/gateway.py:105]: Fixed

File: carberretta/extensions/gateway.py (hunk 3 above)
Comment (neonjonn) [reply carberretta/extensions/gateway.py:124]: Fixed too

File: carberretta/extensions/gateway.py (hunk 4 above)
Comment (neonjonn) [reply carberretta/extensions/gateway.py:74]: Fixed as well, I wasn't aware Hikari did that

**Comment by neonjonn** [comment]
I've applied those changes and copied across the stuff from v1.5.5 :)

**Review by parafoxia — DISMISSED** [review]
These changes look good. I'm gonna add a few additional commits to account for some more differences in Hikari, plus add logging to it. But yeah, your job *should* be done now. Thanks!

**Comment by neonjonn** [comment]
Cool, no problem!

**Review by parafoxia — APPROVED** [review]

**Pull request merged**